File: src/main/kotlin/io/github/ephemient/aoc2019/Day8.kt
package io.github.ephemient.aoc2019

class Day8(private val lines: List<String>, private val width: Int = 25, private val height: Int = 6) {
    fun part1(): Int? = lines.firstOrNull()
        ?.chunked(width * height)
        ?.minBy { it.count { it == '0' } }
        ?.let { it.count { it == '1' } * it.count { it == '2' } }

    fun part2(): String? = lines.firstOrNull()
        ?.map { it - '0' }
        ?.chunked(width * height)
        ?.foldRight<List<Int>, List<Int>?>(null) { s, acc ->
            acc?.zip(s) { a, b -> if (b == 2) a else b } ?: s
        }
        ?.chunked(width)
        ?.joinToString("\n") {
            it.joinToString("", transform = OUTPUT::get)
        }

    companion object {
        private val OUTPUT = listOf("\u2592", "\u2593", "\u2591")
    }
}

File: src/aoc2019/day8.py
import fileinput


def part1(lines, /, width=25, height=6):
    line = lines[0].strip()
    chunk = min((line[i:i + width * height]
                 for i in range(0, len(line), width * height)),
                key=lambda s: s.count('0'))
    return chunk.count('1') * chunk.count('2')


def part2(lines, /, width=25, height=6):
    line = lines[0].strip()
    pixels = ''.join(
        ''.join(stack).lstrip('2')[:1] or '2'
        for stack in zip(*(line[i:i + width * height]
                           for i in range(0, len(line), width * height))))
    trans = str.maketrans('012', '\u2592\u2593\u2591')
    return '\n'.join(pixels[i * width:(i + 1) * width].translate(trans)
                     for i in range(height))


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))
